--- diagrams-core-0.4.api	2012-08-02 21:06:30.220915433 -0700
+++ diagrams-core-0.5.1.api	2012-08-02 21:06:19.330915436 -0700
@@ -1,165 +1,176 @@
 ----------------------------------------
 module Graphics.Rendering.Diagrams
 
-infixl 9 Graphics.Rendering.Diagrams.Bounds.appBounds
-Graphics.Rendering.Diagrams.Bounds.appBounds ::
-  forall v.
-    () =>
-    Graphics.Rendering.Diagrams.Bounds.Bounds v ->
-    v ->
-    Data.VectorSpace.Scalar v
-
-infixl 9 Graphics.Rendering.Diagrams.Bounds.boundary
-Graphics.Rendering.Diagrams.Bounds.boundary ::
-  forall a.
-    (
-     Graphics.Rendering.Diagrams.Bounds.Boundable a
-     ) =>
-    Graphics.Rendering.Diagrams.V.V a ->
-    a ->
-    Graphics.Rendering.Diagrams.Points.Point Graphics.Rendering.Diagrams.V.V a
-
-infixl 9 Graphics.Rendering.Diagrams.Bounds.boundaryFrom
-Graphics.Rendering.Diagrams.Bounds.boundaryFrom ::
-  forall a.
-    (
-     Graphics.Rendering.Diagrams.Bounds.Boundable a
-     ) =>
-    Graphics.Rendering.Diagrams.Points.Point Graphics.Rendering.Diagrams.V.V a ->
-    Graphics.Rendering.Diagrams.V.V a ->
-    a ->
-    Graphics.Rendering.Diagrams.Points.Point Graphics.Rendering.Diagrams.V.V a
-
-infixl 9 Graphics.Rendering.Diagrams.Bounds.boundaryV
-Graphics.Rendering.Diagrams.Bounds.boundaryV ::
-  forall a.
-    (
-     Graphics.Rendering.Diagrams.Bounds.Boundable a
-     ) =>
-    Graphics.Rendering.Diagrams.V.V a ->
-    a ->
-    Graphics.Rendering.Diagrams.V.V a
-
-infixl 9 Graphics.Rendering.Diagrams.Bounds.diameter
-Graphics.Rendering.Diagrams.Bounds.diameter ::
-  forall a.
-    (
-     Graphics.Rendering.Diagrams.Bounds.Boundable a
-     ) =>
-    Graphics.Rendering.Diagrams.V.V a ->
-    a ->
-    Data.VectorSpace.Scalar Graphics.Rendering.Diagrams.V.V a
-
-infixl 9 Graphics.Rendering.Diagrams.Bounds.radius
-Graphics.Rendering.Diagrams.Bounds.radius ::
-  forall a.
-    (
-     Graphics.Rendering.Diagrams.Bounds.Boundable a
-     ) =>
-    Graphics.Rendering.Diagrams.V.V a ->
-    a ->
-    Data.VectorSpace.Scalar Graphics.Rendering.Diagrams.V.V a
-
 infixl 6 Graphics.Rendering.Diagrams.Core.atop
 Graphics.Rendering.Diagrams.Core.atop ::
   forall v m b.
     (
-     Data.Monoid.Monoid m
-     ,
-     Graphics.Rendering.Diagrams.Bounds.OrderedField Data.VectorSpace.Scalar v
+     Graphics.Rendering.Diagrams.Envelope.OrderedField Data.VectorSpace.Scalar v
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
+     Data.Semigroup.Semigroup m
+     ,
      Data.VectorSpace.InnerSpace v
      ) =>
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m
+
+infixl 9 Graphics.Rendering.Diagrams.Core.clearValue
+Graphics.Rendering.Diagrams.Core.clearValue ::
+  forall b v m.
+    () =>
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v Data.Monoid.Any
 
-infixl 9 Graphics.Rendering.Diagrams.Core.bounds
-Graphics.Rendering.Diagrams.Core.bounds ::
+infixl 9 Graphics.Rendering.Diagrams.Core.envelope
+Graphics.Rendering.Diagrams.Core.envelope ::
   forall v b m.
     (
-     Graphics.Rendering.Diagrams.Bounds.OrderedField Data.VectorSpace.Scalar v
+     Graphics.Rendering.Diagrams.Envelope.OrderedField Data.VectorSpace.Scalar v
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
      Data.VectorSpace.InnerSpace v
      ) =>
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    Graphics.Rendering.Diagrams.Bounds.Bounds v
-
-infixl 9 Graphics.Rendering.Diagrams.Core.clearValue
-Graphics.Rendering.Diagrams.Core.clearValue ::
-  forall b v m.
-    () =>
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v Data.Monoid.Any
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Envelope.Envelope v
 
 infixl 9 Graphics.Rendering.Diagrams.Core.freeze
 Graphics.Rendering.Diagrams.Core.freeze ::
   forall v b m.
     (
-     Data.Monoid.Monoid m
+     Graphics.Rendering.Diagrams.Envelope.OrderedField Data.VectorSpace.Scalar v
+     ,
+     Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
-     Graphics.Rendering.Diagrams.Bounds.OrderedField Data.VectorSpace.Scalar v
+     Data.VectorSpace.InnerSpace v
+     ) =>
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m
+
+infixl 9 Graphics.Rendering.Diagrams.Core.fromNames
+Graphics.Rendering.Diagrams.Core.fromNames ::
+  forall a b v m.
+    (
+     Graphics.Rendering.Diagrams.Names.IsName a
+     ) =>
+    [(a, Graphics.Rendering.Diagrams.Core.Subdiagram b v m)] ->
+    Graphics.Rendering.Diagrams.Core.SubMap b v m
+
+infixl 9 Graphics.Rendering.Diagrams.Core.getSub
+Graphics.Rendering.Diagrams.Core.getSub ::
+  forall v b m.
+    (
+     GHC.Float.Floating Data.VectorSpace.Scalar v
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
+     Data.AdditiveGroup.AdditiveGroup Data.VectorSpace.Scalar v
+     ,
      Data.VectorSpace.InnerSpace v
      ) =>
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m
+    Graphics.Rendering.Diagrams.Core.Subdiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m
+
+infixl 9 Graphics.Rendering.Diagrams.Core.location
+Graphics.Rendering.Diagrams.Core.location ::
+  forall v b m.
+    (
+     Graphics.Rendering.Diagrams.Transform.HasLinearMap v
+     ) =>
+    Graphics.Rendering.Diagrams.Core.Subdiagram b v m ->
+    Data.AffineSpace.Point.Point v
 
-infixl 9 Graphics.Rendering.Diagrams.Core.mkAD
-Graphics.Rendering.Diagrams.Core.mkAD ::
+infixl 9 Graphics.Rendering.Diagrams.Core.lookupSub
+Graphics.Rendering.Diagrams.Core.lookupSub ::
+  forall n b v m.
+    (
+     Graphics.Rendering.Diagrams.Names.IsName n
+     ) =>
+    n ->
+    Graphics.Rendering.Diagrams.Core.SubMap b v m ->
+    Data.Maybe.Maybe [Graphics.Rendering.Diagrams.Core.Subdiagram b v m]
+
+infixl 9 Graphics.Rendering.Diagrams.Core.mkQD
+Graphics.Rendering.Diagrams.Core.mkQD ::
   forall b v m.
     () =>
     Graphics.Rendering.Diagrams.Core.Prim b v ->
-    Graphics.Rendering.Diagrams.Bounds.Bounds v ->
-    Graphics.Rendering.Diagrams.Names.NameMap v ->
+    Graphics.Rendering.Diagrams.Envelope.Envelope v ->
+    Graphics.Rendering.Diagrams.Trace.Trace v ->
+    Graphics.Rendering.Diagrams.Core.SubMap b v m ->
     Graphics.Rendering.Diagrams.Query.Query v m ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m
+
+infixl 9 Graphics.Rendering.Diagrams.Core.mkSubdiagram
+Graphics.Rendering.Diagrams.Core.mkSubdiagram ::
+  forall b v m.
+    () =>
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.Subdiagram b v m
 
 infixl 9 Graphics.Rendering.Diagrams.Core.namePoint
 Graphics.Rendering.Diagrams.Core.namePoint ::
   forall v b n m.
     (
-     Data.Monoid.Monoid m
+     Graphics.Rendering.Diagrams.Core.Backend b v
+     ,
+     Graphics.Rendering.Diagrams.Envelope.OrderedField Data.VectorSpace.Scalar v
      ,
-     Graphics.Rendering.Diagrams.Bounds.OrderedField Data.VectorSpace.Scalar v
+     Graphics.Rendering.Diagrams.Names.IsName n
+     ,
+     Graphics.Rendering.Diagrams.Transform.HasLinearMap v
+     ,
+     Data.Monoid.WithSemigroup.Monoid' m
+     ,
+     Data.VectorSpace.InnerSpace v
+     ) =>
+    (Graphics.Rendering.Diagrams.Core.QDiagram b v m -> Data.AffineSpace.Point.Point v) ->
+    n ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m
+
+infixl 9 Graphics.Rendering.Diagrams.Core.nameSub
+Graphics.Rendering.Diagrams.Core.nameSub ::
+  forall v b n m.
+    (
+     Graphics.Rendering.Diagrams.Envelope.OrderedField Data.VectorSpace.Scalar v
      ,
      Graphics.Rendering.Diagrams.Names.IsName n
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
+     Data.Semigroup.Semigroup m
+     ,
      Data.VectorSpace.InnerSpace v
      ) =>
-    (Graphics.Rendering.Diagrams.Core.AnnDiagram b v m -> (Graphics.Rendering.Diagrams.Points.Point v, Graphics.Rendering.Diagrams.Bounds.Bounds v)) ->
+    (Graphics.Rendering.Diagrams.Core.QDiagram b v m -> Graphics.Rendering.Diagrams.Core.Subdiagram b v m) ->
     n ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m
 
 infixl 9 Graphics.Rendering.Diagrams.Core.named
 Graphics.Rendering.Diagrams.Core.named ::
   forall v b n m.
     (
-     Data.Monoid.Monoid m
-     ,
-     Graphics.Rendering.Diagrams.Bounds.OrderedField Data.VectorSpace.Scalar v
+     Graphics.Rendering.Diagrams.Envelope.OrderedField Data.VectorSpace.Scalar v
      ,
      Graphics.Rendering.Diagrams.Names.IsName n
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
+     Data.Semigroup.Semigroup m
+     ,
      Data.VectorSpace.InnerSpace v
      ) =>
     n ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m
 
 infixl 9 Graphics.Rendering.Diagrams.Core.names
 Graphics.Rendering.Diagrams.Core.names ::
   forall v b m.
     (
      GHC.Float.Floating Data.VectorSpace.Scalar v
@@ -167,14 +178,14 @@
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
      Data.AdditiveGroup.AdditiveGroup Data.VectorSpace.Scalar v
      ,
      Data.VectorSpace.InnerSpace v
      ) =>
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    Graphics.Rendering.Diagrams.Names.NameMap v
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.SubMap b v m
 
 infixl 9 Graphics.Rendering.Diagrams.Core.nullPrim
 Graphics.Rendering.Diagrams.Core.nullPrim ::
   forall v b.
     (
      Data.Monoid.Monoid Graphics.Rendering.Diagrams.Core.Render b v
@@ -182,84 +193,144 @@
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ) =>
     Graphics.Rendering.Diagrams.Core.Prim b v
 
 infixl 9 Graphics.Rendering.Diagrams.Core.prims
 Graphics.Rendering.Diagrams.Core.prims ::
-  forall v m b.
+  forall v b m.
     (
-     Data.Monoid.Monoid m
-     ,
-     Graphics.Rendering.Diagrams.Bounds.OrderedField Data.VectorSpace.Scalar v
+     Graphics.Rendering.Diagrams.Envelope.OrderedField Data.VectorSpace.Scalar v
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
      Data.VectorSpace.InnerSpace v
      ) =>
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    [(Graphics.Rendering.Diagrams.Core.Prim b v, (Graphics.Rendering.Diagrams.Monoids.Split Graphics.Rendering.Diagrams.Transform.Transformation v, Graphics.Rendering.Diagrams.Style.Style v))]
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    [(Graphics.Rendering.Diagrams.Core.Prim b v, (Data.Monoid.Split.Split Graphics.Rendering.Diagrams.Transform.Transformation v, Graphics.Rendering.Diagrams.Style.Style v))]
 
 infixl 9 Graphics.Rendering.Diagrams.Core.query
 Graphics.Rendering.Diagrams.Core.query ::
   forall v m b.
     (
      Data.Monoid.Monoid m
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ) =>
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
     Graphics.Rendering.Diagrams.Query.Query v m
 
+infixl 9 Graphics.Rendering.Diagrams.Core.rawSub
+Graphics.Rendering.Diagrams.Core.rawSub ::
+  forall b v m.
+    () =>
+    Graphics.Rendering.Diagrams.Core.Subdiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m
+
+infixl 9 Graphics.Rendering.Diagrams.Core.rememberAs
+Graphics.Rendering.Diagrams.Core.rememberAs ::
+  forall a b v m.
+    (
+     Graphics.Rendering.Diagrams.Names.IsName a
+     ) =>
+    a ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.SubMap b v m ->
+    Graphics.Rendering.Diagrams.Core.SubMap b v m
+
 infixl 9 Graphics.Rendering.Diagrams.Core.resetValue
 Graphics.Rendering.Diagrams.Core.resetValue ::
   forall m b v.
     (
      Data.Monoid.Monoid m
      ,
      GHC.Classes.Eq m
      ) =>
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v Data.Monoid.Any
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v Data.Monoid.Any
 
 infixl 9 Graphics.Rendering.Diagrams.Core.sample
 Graphics.Rendering.Diagrams.Core.sample ::
   forall v m b.
     (
      Data.Monoid.Monoid m
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ) =>
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    Graphics.Rendering.Diagrams.Points.Point v ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Data.AffineSpace.Point.Point v ->
     m
 
-infixl 9 Graphics.Rendering.Diagrams.Core.setBounds
-Graphics.Rendering.Diagrams.Core.setBounds ::
+infixl 9 Graphics.Rendering.Diagrams.Core.setEnvelope
+Graphics.Rendering.Diagrams.Core.setEnvelope ::
   forall b v m.
     (
-     Data.Monoid.Monoid m
+     Graphics.Rendering.Diagrams.Envelope.OrderedField Data.VectorSpace.Scalar v
+     ,
+     Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
-     Graphics.Rendering.Diagrams.Bounds.OrderedField Data.VectorSpace.Scalar v
+     Data.Monoid.WithSemigroup.Monoid' m
+     ,
+     Data.VectorSpace.InnerSpace v
+     ) =>
+    Graphics.Rendering.Diagrams.Envelope.Envelope v ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m
+
+infixl 9 Graphics.Rendering.Diagrams.Core.setTrace
+Graphics.Rendering.Diagrams.Core.setTrace ::
+  forall b v m.
+    (
+     Graphics.Rendering.Diagrams.Envelope.OrderedField Data.VectorSpace.Scalar v
+     ,
+     Graphics.Rendering.Diagrams.Transform.HasLinearMap v
+     ,
+     Data.Monoid.WithSemigroup.Monoid' m
+     ,
+     Data.VectorSpace.InnerSpace v
+     ) =>
+    Graphics.Rendering.Diagrams.Trace.Trace v ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m
+
+infixl 9 Graphics.Rendering.Diagrams.Core.subPoint
+Graphics.Rendering.Diagrams.Core.subPoint ::
+  forall v m b.
+    (
+     Graphics.Rendering.Diagrams.Envelope.OrderedField Data.VectorSpace.Scalar v
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
+     Data.Semigroup.Semigroup m
+     ,
      Data.VectorSpace.InnerSpace v
      ) =>
-    Graphics.Rendering.Diagrams.Bounds.Bounds v ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m
+    Data.AffineSpace.Point.Point v ->
+    Graphics.Rendering.Diagrams.Core.Subdiagram b v m
+
+infixl 9 Graphics.Rendering.Diagrams.Core.trace
+Graphics.Rendering.Diagrams.Core.trace ::
+  forall v b m.
+    (
+     Graphics.Rendering.Diagrams.Transform.HasLinearMap v
+     ,
+     GHC.Classes.Ord Data.VectorSpace.Scalar v
+     ,
+     Data.VectorSpace.VectorSpace v
+     ) =>
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Trace.Trace v
 
 infixl 9 Graphics.Rendering.Diagrams.Core.value
 Graphics.Rendering.Diagrams.Core.value ::
   forall m b v.
     (
      Data.Monoid.Monoid m
      ) =>
     m ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v Data.Monoid.Any ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m
+    Graphics.Rendering.Diagrams.Core.QDiagram b v Data.Monoid.Any ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m
 
 infixl 9 Graphics.Rendering.Diagrams.Core.withName
 Graphics.Rendering.Diagrams.Core.withName ::
   forall n v b m.
     (
      GHC.Float.Floating Data.VectorSpace.Scalar v
@@ -270,15 +341,15 @@
      ,
      Data.AdditiveGroup.AdditiveGroup Data.VectorSpace.Scalar v
      ,
      Data.VectorSpace.InnerSpace v
      ) =>
     n ->
-    ((Graphics.Rendering.Diagrams.Points.Point v, Graphics.Rendering.Diagrams.Bounds.Bounds v) -> Graphics.Rendering.Diagrams.Core.AnnDiagram b v m -> Graphics.Rendering.Diagrams.Core.AnnDiagram b v m) ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m
+    (Graphics.Rendering.Diagrams.Core.Subdiagram b v m -> Graphics.Rendering.Diagrams.Core.QDiagram b v m -> Graphics.Rendering.Diagrams.Core.QDiagram b v m) ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m
 
 infixl 9 Graphics.Rendering.Diagrams.Core.withNameAll
 Graphics.Rendering.Diagrams.Core.withNameAll ::
   forall n v b m.
     (
      GHC.Float.Floating Data.VectorSpace.Scalar v
@@ -289,15 +360,15 @@
      ,
      Data.AdditiveGroup.AdditiveGroup Data.VectorSpace.Scalar v
      ,
      Data.VectorSpace.InnerSpace v
      ) =>
     n ->
-    ([(Graphics.Rendering.Diagrams.Points.Point v, Graphics.Rendering.Diagrams.Bounds.Bounds v)] -> Graphics.Rendering.Diagrams.Core.AnnDiagram b v m -> Graphics.Rendering.Diagrams.Core.AnnDiagram b v m) ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m
+    ([Graphics.Rendering.Diagrams.Core.Subdiagram b v m] -> Graphics.Rendering.Diagrams.Core.QDiagram b v m -> Graphics.Rendering.Diagrams.Core.QDiagram b v m) ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m
 
 infixl 9 Graphics.Rendering.Diagrams.Core.withNames
 Graphics.Rendering.Diagrams.Core.withNames ::
   forall n v b m.
     (
      GHC.Float.Floating Data.VectorSpace.Scalar v
@@ -308,15 +379,85 @@
      ,
      Data.AdditiveGroup.AdditiveGroup Data.VectorSpace.Scalar v
      ,
      Data.VectorSpace.InnerSpace v
      ) =>
     [n] ->
-    ([(Graphics.Rendering.Diagrams.Points.Point v, Graphics.Rendering.Diagrams.Bounds.Bounds v)] -> Graphics.Rendering.Diagrams.Core.AnnDiagram b v m -> Graphics.Rendering.Diagrams.Core.AnnDiagram b v m) ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-    Graphics.Rendering.Diagrams.Core.AnnDiagram b v m
+    ([Graphics.Rendering.Diagrams.Core.Subdiagram b v m] -> Graphics.Rendering.Diagrams.Core.QDiagram b v m -> Graphics.Rendering.Diagrams.Core.QDiagram b v m) ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+    Graphics.Rendering.Diagrams.Core.QDiagram b v m
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.appEnvelope
+Graphics.Rendering.Diagrams.Envelope.appEnvelope ::
+  forall v.
+    () =>
+    Graphics.Rendering.Diagrams.Envelope.Envelope v ->
+    Data.Maybe.Maybe (v -> Data.VectorSpace.Scalar v)
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.diameter
+Graphics.Rendering.Diagrams.Envelope.diameter ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Envelope.Enveloped a
+     ) =>
+    Graphics.Rendering.Diagrams.V.V a ->
+    a ->
+    Data.VectorSpace.Scalar Graphics.Rendering.Diagrams.V.V a
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.envelopeP
+Graphics.Rendering.Diagrams.Envelope.envelopeP ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Envelope.Enveloped a
+     ) =>
+    Graphics.Rendering.Diagrams.V.V a ->
+    a ->
+    Data.AffineSpace.Point.Point Graphics.Rendering.Diagrams.V.V a
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.envelopeV
+Graphics.Rendering.Diagrams.Envelope.envelopeV ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Envelope.Enveloped a
+     ) =>
+    Graphics.Rendering.Diagrams.V.V a ->
+    a ->
+    Graphics.Rendering.Diagrams.V.V a
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.inEnvelope
+Graphics.Rendering.Diagrams.Envelope.inEnvelope ::
+  forall v.
+    () =>
+    (Data.Semigroup.Option (v -> Data.Semigroup.Max Data.VectorSpace.Scalar v) -> Data.Semigroup.Option (v -> Data.Semigroup.Max Data.VectorSpace.Scalar v)) ->
+    Graphics.Rendering.Diagrams.Envelope.Envelope v ->
+    Graphics.Rendering.Diagrams.Envelope.Envelope v
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.mkEnvelope
+Graphics.Rendering.Diagrams.Envelope.mkEnvelope ::
+  forall v.
+    () =>
+    (v -> Data.VectorSpace.Scalar v) ->
+    Graphics.Rendering.Diagrams.Envelope.Envelope v
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.onEnvelope
+Graphics.Rendering.Diagrams.Envelope.onEnvelope ::
+  forall v.
+    () =>
+    ((v -> Data.VectorSpace.Scalar v) -> v -> Data.VectorSpace.Scalar v) ->
+    Graphics.Rendering.Diagrams.Envelope.Envelope v ->
+    Graphics.Rendering.Diagrams.Envelope.Envelope v
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.radius
+Graphics.Rendering.Diagrams.Envelope.radius ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Envelope.Enveloped a
+     ) =>
+    Graphics.Rendering.Diagrams.V.V a ->
+    a ->
+    Data.VectorSpace.Scalar Graphics.Rendering.Diagrams.V.V a
 
 infixl 9 Graphics.Rendering.Diagrams.HasOrigin.moveOriginBy
 Graphics.Rendering.Diagrams.HasOrigin.moveOriginBy ::
   forall t.
     (
      Graphics.Rendering.Diagrams.HasOrigin.HasOrigin t
@@ -328,106 +469,57 @@
 infixl 9 Graphics.Rendering.Diagrams.HasOrigin.moveTo
 Graphics.Rendering.Diagrams.HasOrigin.moveTo ::
   forall t.
     (
      Graphics.Rendering.Diagrams.HasOrigin.HasOrigin t
      ) =>
-    Graphics.Rendering.Diagrams.Points.Point Graphics.Rendering.Diagrams.V.V t ->
+    Data.AffineSpace.Point.Point Graphics.Rendering.Diagrams.V.V t ->
     t ->
     t
 
 infixl 9 Graphics.Rendering.Diagrams.HasOrigin.place
 Graphics.Rendering.Diagrams.HasOrigin.place ::
   forall t.
     (
      Graphics.Rendering.Diagrams.HasOrigin.HasOrigin t
      ) =>
     t ->
-    Graphics.Rendering.Diagrams.Points.Point Graphics.Rendering.Diagrams.V.V t ->
+    Data.AffineSpace.Point.Point Graphics.Rendering.Diagrams.V.V t ->
     t
 
+infixl 9 Graphics.Rendering.Diagrams.Juxtapose.juxtaposeDefault
+Graphics.Rendering.Diagrams.Juxtapose.juxtaposeDefault ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Envelope.Enveloped a
+     ,
+     Graphics.Rendering.Diagrams.HasOrigin.HasOrigin a
+     ) =>
+    Graphics.Rendering.Diagrams.V.V a ->
+    a ->
+    a ->
+    a
+
 infixr 5 (Graphics.Rendering.Diagrams.Names..>)
 (Graphics.Rendering.Diagrams.Names..>) ::
   forall a1 a2.
     (
      Graphics.Rendering.Diagrams.Names.IsName a1
      ,
      Graphics.Rendering.Diagrams.Names.IsName a2
      ) =>
     a1 ->
     a2 ->
     Graphics.Rendering.Diagrams.Names.Name
 
-infixl 9 Graphics.Rendering.Diagrams.Names.fromNames
-Graphics.Rendering.Diagrams.Names.fromNames ::
-  forall v a.
-    (
-     Graphics.Rendering.Diagrams.Names.IsName a
-     ,
-     GHC.Classes.Ord Data.VectorSpace.Scalar v
-     ,
-     Data.AdditiveGroup.AdditiveGroup Data.VectorSpace.Scalar v
-     ) =>
-    [(a, Graphics.Rendering.Diagrams.Points.Point v)] ->
-    Graphics.Rendering.Diagrams.Names.NameMap v
-
-infixl 9 Graphics.Rendering.Diagrams.Names.fromNamesB
-Graphics.Rendering.Diagrams.Names.fromNamesB ::
-  forall a v.
-    (
-     Graphics.Rendering.Diagrams.Names.IsName a
-     ) =>
-    [(a, (Graphics.Rendering.Diagrams.Points.Point v, Graphics.Rendering.Diagrams.Bounds.Bounds v))] ->
-    Graphics.Rendering.Diagrams.Names.NameMap v
-
-infixl 9 Graphics.Rendering.Diagrams.Names.lookupN
-Graphics.Rendering.Diagrams.Names.lookupN ::
-  forall n v.
-    (
-     Graphics.Rendering.Diagrams.Names.IsName n
-     ) =>
-    n ->
-    Graphics.Rendering.Diagrams.Names.NameMap v ->
-    Data.Maybe.Maybe [(Graphics.Rendering.Diagrams.Points.Point v, Graphics.Rendering.Diagrams.Bounds.Bounds v)]
-
-infixl 9 Graphics.Rendering.Diagrams.Names.rememberAs
-Graphics.Rendering.Diagrams.Names.rememberAs ::
-  forall a v.
-    (
-     Graphics.Rendering.Diagrams.Names.IsName a
-     ) =>
-    a ->
-    Graphics.Rendering.Diagrams.Points.Point v ->
-    Graphics.Rendering.Diagrams.Bounds.Bounds v ->
-    Graphics.Rendering.Diagrams.Names.NameMap v ->
-    Graphics.Rendering.Diagrams.Names.NameMap v
-
-infixl 9 (Graphics.Rendering.Diagrams.Points.*.)
-(Graphics.Rendering.Diagrams.Points.*.) ::
-  forall v.
-    (
-     Data.VectorSpace.VectorSpace v
-     ) =>
-    Data.VectorSpace.Scalar v ->
-    Graphics.Rendering.Diagrams.Points.Point v ->
-    Graphics.Rendering.Diagrams.Points.Point v
-
-infixl 9 Graphics.Rendering.Diagrams.Points.origin
-Graphics.Rendering.Diagrams.Points.origin ::
-  forall v.
-    (
-     Data.AdditiveGroup.AdditiveGroup v
-     ) =>
-    Graphics.Rendering.Diagrams.Points.Point v
-
 infixl 9 Graphics.Rendering.Diagrams.Query.runQuery
 Graphics.Rendering.Diagrams.Query.runQuery ::
   forall v m.
     () =>
     Graphics.Rendering.Diagrams.Query.Query v m ->
-    Graphics.Rendering.Diagrams.Points.Point v ->
+    Data.AffineSpace.Point.Point v ->
     m
 
 infixl 9 Graphics.Rendering.Diagrams.Style.applyAttr
 Graphics.Rendering.Diagrams.Style.applyAttr ::
   forall a d.
     (
@@ -502,12 +594,80 @@
     (
      Graphics.Rendering.Diagrams.Style.AttributeClass a
      ) =>
     Graphics.Rendering.Diagrams.Style.Attribute v ->
     Data.Maybe.Maybe a
 
+infixl 9 Graphics.Rendering.Diagrams.Trace.appTrace
+Graphics.Rendering.Diagrams.Trace.appTrace ::
+  forall v.
+    () =>
+    Graphics.Rendering.Diagrams.Trace.Trace v ->
+    Data.AffineSpace.Point.Point v ->
+    v ->
+    Data.Monoid.PosInf.PosInf Data.VectorSpace.Scalar v
+
+infixl 9 Graphics.Rendering.Diagrams.Trace.inTrace
+Graphics.Rendering.Diagrams.Trace.inTrace ::
+  forall v.
+    () =>
+    ((Data.AffineSpace.Point.Point v -> v -> Data.Monoid.PosInf.PosInf Data.VectorSpace.Scalar v) -> Data.AffineSpace.Point.Point v -> v -> Data.Monoid.PosInf.PosInf Data.VectorSpace.Scalar v) ->
+    Graphics.Rendering.Diagrams.Trace.Trace v ->
+    Graphics.Rendering.Diagrams.Trace.Trace v
+
+infixl 9 Graphics.Rendering.Diagrams.Trace.maxTraceP
+Graphics.Rendering.Diagrams.Trace.maxTraceP ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Trace.Traced a
+     ) =>
+    Data.AffineSpace.Point.Point Graphics.Rendering.Diagrams.V.V a ->
+    Graphics.Rendering.Diagrams.V.V a ->
+    a ->
+    Data.Maybe.Maybe Data.AffineSpace.Point.Point Graphics.Rendering.Diagrams.V.V a
+
+infixl 9 Graphics.Rendering.Diagrams.Trace.maxTraceV
+Graphics.Rendering.Diagrams.Trace.maxTraceV ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Trace.Traced a
+     ) =>
+    Data.AffineSpace.Point.Point Graphics.Rendering.Diagrams.V.V a ->
+    Graphics.Rendering.Diagrams.V.V a ->
+    a ->
+    Data.Maybe.Maybe Graphics.Rendering.Diagrams.V.V a
+
+infixl 9 Graphics.Rendering.Diagrams.Trace.mkTrace
+Graphics.Rendering.Diagrams.Trace.mkTrace ::
+  forall v.
+    () =>
+    (Data.AffineSpace.Point.Point v -> v -> Data.Monoid.PosInf.PosInf Data.VectorSpace.Scalar v) ->
+    Graphics.Rendering.Diagrams.Trace.Trace v
+
+infixl 9 Graphics.Rendering.Diagrams.Trace.traceP
+Graphics.Rendering.Diagrams.Trace.traceP ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Trace.Traced a
+     ) =>
+    Data.AffineSpace.Point.Point Graphics.Rendering.Diagrams.V.V a ->
+    Graphics.Rendering.Diagrams.V.V a ->
+    a ->
+    Data.Maybe.Maybe Data.AffineSpace.Point.Point Graphics.Rendering.Diagrams.V.V a
+
+infixl 9 Graphics.Rendering.Diagrams.Trace.traceV
+Graphics.Rendering.Diagrams.Trace.traceV ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Trace.Traced a
+     ) =>
+    Data.AffineSpace.Point.Point Graphics.Rendering.Diagrams.V.V a ->
+    Graphics.Rendering.Diagrams.V.V a ->
+    a ->
+    Data.Maybe.Maybe Graphics.Rendering.Diagrams.V.V a
+
 infixl 9 (Graphics.Rendering.Diagrams.Transform.<->)
 (Graphics.Rendering.Diagrams.Transform.<->) ::
   forall u v.
     (
      Graphics.Rendering.Diagrams.Transform.HasLinearMap u
      ,
@@ -571,14 +731,14 @@
 Graphics.Rendering.Diagrams.Transform.papply ::
   forall v.
     (
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ) =>
     Graphics.Rendering.Diagrams.Transform.Transformation v ->
-    Graphics.Rendering.Diagrams.Points.Point v ->
-    Graphics.Rendering.Diagrams.Points.Point v
+    Data.AffineSpace.Point.Point v ->
+    Data.AffineSpace.Point.Point v
 
 infixl 9 Graphics.Rendering.Diagrams.Transform.scale
 Graphics.Rendering.Diagrams.Transform.scale ::
   forall t.
     (
      GHC.Real.Fractional Data.VectorSpace.Scalar Graphics.Rendering.Diagrams.V.V t
@@ -641,60 +801,29 @@
 Graphics.Rendering.Diagrams.Transform.unTransInv ::
   forall t.
     () =>
     Graphics.Rendering.Diagrams.Transform.TransInv t ->
     t
 
-infixl 9 Graphics.Rendering.Diagrams.Util.withLength
-Graphics.Rendering.Diagrams.Util.withLength ::
+infixl 9 (Data.AffineSpace.Point.*.)
+(Data.AffineSpace.Point.*.) ::
   forall v.
     (
-     GHC.Float.Floating Data.VectorSpace.Scalar v
-     ,
-     Data.VectorSpace.InnerSpace v
+     Data.VectorSpace.VectorSpace v
      ) =>
     Data.VectorSpace.Scalar v ->
-    v ->
-    v
-
-class (
-       Graphics.Rendering.Diagrams.Bounds.OrderedField Data.VectorSpace.Scalar Graphics.Rendering.Diagrams.V.V b
-       ,
-       Data.VectorSpace.InnerSpace Graphics.Rendering.Diagrams.V.V b
-       ) =>
-  Graphics.Rendering.Diagrams.Bounds.Boundable b where
-    Graphics.Rendering.Diagrams.Bounds.getBounds ::
-      forall b.
-        (
-         Graphics.Rendering.Diagrams.Bounds.Boundable b
-         ) =>
-        b ->
-        Graphics.Rendering.Diagrams.Bounds.Bounds Graphics.Rendering.Diagrams.V.V b
+    Data.AffineSpace.Point.Point v ->
+    Data.AffineSpace.Point.Point v
 
-newtype () =>
-  Bounds' v =
-    Bounds
-      Graphics.Rendering.Diagrams.Bounds.appBounds :: (v -> Data.VectorSpace.Scalar v)
-  deriving ()
-
-class (
-       GHC.Float.Floating s
-       ,
-       GHC.Real.Fractional s
-       ,
-       GHC.Classes.Ord s
-       ,
-       Data.AdditiveGroup.AdditiveGroup s
-       ) =>
-  Graphics.Rendering.Diagrams.Bounds.OrderedField s where
-
-newtype () =>
-  Graphics.Rendering.Diagrams.Core.AnnDiagram b v m =
-    Graphics.Rendering.Diagrams.Core.AD
-      Graphics.Rendering.Diagrams.Core.unAD :: Graphics.Rendering.Diagrams.UDTree.UDTree Graphics.Rendering.Diagrams.Core.UpAnnots v m Graphics.Rendering.Diagrams.Core.DownAnnots v Graphics.Rendering.Diagrams.Core.Prim b v
-  deriving ()
+infixl 9 Data.AffineSpace.Point.origin
+Data.AffineSpace.Point.origin ::
+  forall v.
+    (
+     Data.AdditiveGroup.AdditiveGroup v
+     ) =>
+    Data.AffineSpace.Point.Point v
 
 class (
        Data.Monoid.Monoid Graphics.Rendering.Diagrams.Core.Render b v
        ,
        Graphics.Rendering.Diagrams.Transform.HasLinearMap v
        ) =>
@@ -722,37 +851,39 @@
       forall b v.
         (
          Graphics.Rendering.Diagrams.Core.Backend b v
          ) =>
         forall m.
           (
-           Data.Monoid.Monoid m
+           Data.Monoid.WithSemigroup.Monoid' m
            ) =>
           b ->
           Graphics.Rendering.Diagrams.Core.Options b v ->
-          Graphics.Rendering.Diagrams.Core.AnnDiagram b v m ->
-          Graphics.Rendering.Diagrams.Core.AnnDiagram b v m
+          Graphics.Rendering.Diagrams.Core.QDiagram b v m ->
+          (Graphics.Rendering.Diagrams.Core.Options b v, Graphics.Rendering.Diagrams.Core.QDiagram b v m)
     Graphics.Rendering.Diagrams.Core.renderDia ::
       forall b v.
         (
          Graphics.Rendering.Diagrams.Core.Backend b v
          ) =>
         forall m'.
           (
-           Data.Monoid.Monoid m'
+           Graphics.Rendering.Diagrams.Envelope.OrderedField Data.VectorSpace.Scalar v
            ,
-           Graphics.Rendering.Diagrams.Bounds.OrderedField Data.VectorSpace.Scalar v
+           Data.Monoid.WithSemigroup.Monoid' m'
            ,
            Data.VectorSpace.InnerSpace v
            ) =>
           b ->
           Graphics.Rendering.Diagrams.Core.Options b v ->
-          Graphics.Rendering.Diagrams.Core.AnnDiagram b v m' ->
+          Graphics.Rendering.Diagrams.Core.QDiagram b v m' ->
           Graphics.Rendering.Diagrams.Core.Result b v
 
-type Graphics.Rendering.Diagrams.Core.Diagram b v = Graphics.Rendering.Diagrams.Core.AnnDiagram b v Data.Monoid.Any
+type Graphics.Rendering.Diagrams.Core.D v = Graphics.Rendering.Diagrams.Core.Diagram Graphics.Rendering.Diagrams.Core.NullBackend v
+
+type Graphics.Rendering.Diagrams.Core.Diagram b v = Graphics.Rendering.Diagrams.Core.QDiagram b v Data.Monoid.Any
 
 class (
        Graphics.Rendering.Diagrams.Core.Backend b v
        ) =>
   Graphics.Rendering.Diagrams.Core.MultiBackend b v where
     Graphics.Rendering.Diagrams.Core.renderDias ::
@@ -761,28 +892,38 @@
          Graphics.Rendering.Diagrams.Core.MultiBackend b v
          ) =>
         forall m.
           () =>
           b ->
           Graphics.Rendering.Diagrams.Core.Options b v ->
-          [Graphics.Rendering.Diagrams.Core.AnnDiagram b v m] ->
+          [Graphics.Rendering.Diagrams.Core.QDiagram b v m] ->
           Graphics.Rendering.Diagrams.Core.Result b v
 
+data () =>
+  Graphics.Rendering.Diagrams.Core.NullBackend 
+  deriving ()
+
 data family Graphics.Rendering.Diagrams.Core.Options b v
   :: * -> * -> *
 
 data () =>
   Prim' b v =
-    forall t. 
+    forall p. 
     (
-     Graphics.Rendering.Diagrams.Core.Renderable t b
+     Graphics.Rendering.Diagrams.Core.Renderable p b
      ,
-     v ~ Graphics.Rendering.Diagrams.V.V t
+     v ~ Graphics.Rendering.Diagrams.V.V p
      ) =>
       Prim
-        t
+        p
+  deriving ()
+
+newtype () =>
+  Graphics.Rendering.Diagrams.Core.QDiagram b v m =
+    Graphics.Rendering.Diagrams.Core.QD
+      Graphics.Rendering.Diagrams.Core.unQD :: Data.Tree.DUBL.DUBLTree Graphics.Rendering.Diagrams.Core.DownAnnots v Graphics.Rendering.Diagrams.Core.UpAnnots b v m () Graphics.Rendering.Diagrams.Core.Prim b v
   deriving ()
 
 data family Graphics.Rendering.Diagrams.Core.Render b v
   :: * -> * -> *
 
 class (
@@ -798,25 +939,81 @@
         t ->
         Graphics.Rendering.Diagrams.Core.Render b Graphics.Rendering.Diagrams.V.V t
 
 type family Graphics.Rendering.Diagrams.Core.Result b v
   :: * -> * -> *
 
+newtype () =>
+  Graphics.Rendering.Diagrams.Core.SubMap b v m =
+    Graphics.Rendering.Diagrams.Core.SubMap
+      Data.Map.Map Graphics.Rendering.Diagrams.Names.Name [Graphics.Rendering.Diagrams.Core.Subdiagram b v m]
+  deriving ()
+
+data () =>
+  Graphics.Rendering.Diagrams.Core.Subdiagram b v m =
+    Graphics.Rendering.Diagrams.Core.Subdiagram
+      Graphics.Rendering.Diagrams.Core.QDiagram b v m
+      Graphics.Rendering.Diagrams.Core.DownAnnots v
+  deriving ()
+
+newtype () =>
+  Envelope' v =
+    Envelope
+      Graphics.Rendering.Diagrams.Envelope.unEnvelope :: Data.Semigroup.Option (v -> Data.Semigroup.Max Data.VectorSpace.Scalar v)
+  deriving ()
+
+class (
+       Graphics.Rendering.Diagrams.Envelope.OrderedField Data.VectorSpace.Scalar Graphics.Rendering.Diagrams.V.V a
+       ,
+       Data.VectorSpace.InnerSpace Graphics.Rendering.Diagrams.V.V a
+       ) =>
+  Graphics.Rendering.Diagrams.Envelope.Enveloped a where
+    Graphics.Rendering.Diagrams.Envelope.getEnvelope ::
+      forall a.
+        (
+         Graphics.Rendering.Diagrams.Envelope.Enveloped a
+         ) =>
+        a ->
+        Graphics.Rendering.Diagrams.Envelope.Envelope Graphics.Rendering.Diagrams.V.V a
+
+class (
+       GHC.Float.Floating s
+       ,
+       GHC.Real.Fractional s
+       ,
+       GHC.Classes.Ord s
+       ,
+       Data.AdditiveGroup.AdditiveGroup s
+       ) =>
+  Graphics.Rendering.Diagrams.Envelope.OrderedField s where
+
 class (
        Data.VectorSpace.VectorSpace Graphics.Rendering.Diagrams.V.V t
        ) =>
   Graphics.Rendering.Diagrams.HasOrigin.HasOrigin t where
     Graphics.Rendering.Diagrams.HasOrigin.moveOriginTo ::
       forall t.
         (
          Graphics.Rendering.Diagrams.HasOrigin.HasOrigin t
          ) =>
-        Graphics.Rendering.Diagrams.Points.Point Graphics.Rendering.Diagrams.V.V t ->
+        Data.AffineSpace.Point.Point Graphics.Rendering.Diagrams.V.V t ->
         t ->
         t
 
+class () =>
+  Graphics.Rendering.Diagrams.Juxtapose.Juxtaposable a where
+    Graphics.Rendering.Diagrams.Juxtapose.juxtapose ::
+      forall a.
+        (
+         Graphics.Rendering.Diagrams.Juxtapose.Juxtaposable a
+         ) =>
+        Graphics.Rendering.Diagrams.V.V a ->
+        a ->
+        a ->
+        a
+
 data () =>
   Graphics.Rendering.Diagrams.Names.AName  =
     forall a. 
     (
      Data.Typeable.Internal.Typeable a
      ,
@@ -847,18 +1044,12 @@
 newtype () =>
   Graphics.Rendering.Diagrams.Names.Name  =
     Graphics.Rendering.Diagrams.Names.Name
       [Graphics.Rendering.Diagrams.Names.AName]
   deriving ()
 
-newtype () =>
-  Graphics.Rendering.Diagrams.Names.NameMap v =
-    Graphics.Rendering.Diagrams.Names.NameMap
-      Data.Map.Map Graphics.Rendering.Diagrams.Names.Name [(Graphics.Rendering.Diagrams.Points.Point v, Graphics.Rendering.Diagrams.Transform.TransInv Graphics.Rendering.Diagrams.Bounds.Bounds v)]
-  deriving ()
-
 class () =>
   Graphics.Rendering.Diagrams.Names.Qualifiable q where
     (Graphics.Rendering.Diagrams.Names.|>) ::
       forall q.
         (
          Graphics.Rendering.Diagrams.Names.Qualifiable q
@@ -869,21 +1060,15 @@
            ) =>
           a ->
           q ->
           q
 
 newtype () =>
-  Graphics.Rendering.Diagrams.Points.Point v =
-    Graphics.Rendering.Diagrams.Points.P
-      v
-  deriving ()
-
-newtype () =>
-  Graphics.Rendering.Diagrams.Query.Query v m =
-    Graphics.Rendering.Diagrams.Query.Query
-      Graphics.Rendering.Diagrams.Query.runQuery :: (Graphics.Rendering.Diagrams.Points.Point v -> m)
+  Query' v m =
+    Query
+      Graphics.Rendering.Diagrams.Query.runQuery :: (Data.AffineSpace.Point.Point v -> m)
   deriving ()
 
 data () =>
   Graphics.Rendering.Diagrams.Style.Attribute v =
     forall a. 
     (
@@ -924,12 +1109,32 @@
 newtype () =>
   Graphics.Rendering.Diagrams.Style.Style v =
     Graphics.Rendering.Diagrams.Style.Style
       Data.Map.Map GHC.Base.String Graphics.Rendering.Diagrams.Style.Attribute v
   deriving ()
 
+newtype () =>
+  Trace' v =
+    Trace
+      Graphics.Rendering.Diagrams.Trace.appTrace :: (Data.AffineSpace.Point.Point v -> v -> Data.Monoid.PosInf.PosInf Data.VectorSpace.Scalar v)
+  deriving ()
+
+class (
+       GHC.Classes.Ord Data.VectorSpace.Scalar Graphics.Rendering.Diagrams.V.V a
+       ,
+       Data.VectorSpace.VectorSpace Graphics.Rendering.Diagrams.V.V a
+       ) =>
+  Graphics.Rendering.Diagrams.Trace.Traced a where
+    Graphics.Rendering.Diagrams.Trace.getTrace ::
+      forall a.
+        (
+         Graphics.Rendering.Diagrams.Trace.Traced a
+         ) =>
+        a ->
+        Graphics.Rendering.Diagrams.Trace.Trace Graphics.Rendering.Diagrams.V.V a
+
 data () =>
   (:-:') u v =
     (:-:)
       (Data.LinearMap.:-*) u v
       (Data.LinearMap.:-*) v u
   deriving ()
@@ -970,6 +1175,19 @@
       v
   deriving ()
 
 type family Graphics.Rendering.Diagrams.V.V a
   :: * -> *
 
+class (
+       Data.Monoid.Monoid m
+       ,
+       Data.Semigroup.Semigroup m
+       ) =>
+  Data.Monoid.WithSemigroup.Monoid' m where
+
+newtype () =>
+  Data.AffineSpace.Point.Point v =
+    Data.AffineSpace.Point.P
+      v
+  deriving ()
+
