--- diagrams-core-0.4.api	2012-07-22 16:51:41.898382792 -0700
+++ diagrams-core-0.5.1.api	2012-07-22 15:39:59.428383564 -0700
@@ -2,6 +2,18 @@
 module Graphics.Rendering.Diagrams
 
 class () =>
+  Graphics.Rendering.Diagrams.Juxtapose.Juxtaposable a where
+    Graphics.Rendering.Diagrams.Juxtapose.juxtapose ::
+      forall a.
+        (
+         Graphics.Rendering.Diagrams.Juxtapose.Juxtaposable a
+         ) =>
+        a ->
+        a ->
+        a ->
+        a
+
+class () =>
   Graphics.Rendering.Diagrams.Names.Qualifiable q where
     (Graphics.Rendering.Diagrams.Names.|>) ::
       forall q.
@@ -59,7 +71,7 @@
        ,
        Data.AdditiveGroup.AdditiveGroup s
        ) =>
-  Graphics.Rendering.Diagrams.Bounds.OrderedField s where
+  Graphics.Rendering.Diagrams.Envelope.OrderedField s where
 
 class (
        Graphics.Rendering.Diagrams.Core.Backend b v
@@ -109,12 +121,12 @@
          ) =>
         forall m.
           (
-           Data.Monoid.Monoid m
+           Data.Monoid.WithSemigroup.Monoid' m
            ) =>
           b ->
           b v ->
           b v m ->
-          b v m
+          (b v, b v m)
     Graphics.Rendering.Diagrams.Core.renderDia ::
       forall b v.
         (
@@ -122,9 +134,9 @@
          ) =>
         forall m'.
           (
-           Data.Monoid.Monoid m'
+           Graphics.Rendering.Diagrams.Envelope.OrderedField v
            ,
-           Graphics.Rendering.Diagrams.Bounds.OrderedField v
+           Data.Monoid.WithSemigroup.Monoid' m'
            ,
            Data.VectorSpace.InnerSpace v
            ) =>
@@ -160,6 +172,27 @@
         b t
 
 class (
+       GHC.Classes.Ord a
+       ,
+       Data.VectorSpace.VectorSpace a
+       ) =>
+  Graphics.Rendering.Diagrams.Trace.Traced a where
+    Graphics.Rendering.Diagrams.Trace.getTrace ::
+      forall a.
+        (
+         Graphics.Rendering.Diagrams.Trace.Traced a
+         ) =>
+        a ->
+        a
+
+class (
+       Data.Monoid.Monoid m
+       ,
+       Data.Semigroup.Semigroup m
+       ) =>
+  Data.Monoid.WithSemigroup.Monoid' m where
+
+class (
        Data.MemoTrie.HasTrie v
        ,
        Data.Basis.HasBasis v
@@ -169,18 +202,18 @@
   Graphics.Rendering.Diagrams.Transform.HasLinearMap v where
 
 class (
-       Graphics.Rendering.Diagrams.Bounds.OrderedField b
+       Graphics.Rendering.Diagrams.Envelope.OrderedField a
        ,
-       Data.VectorSpace.InnerSpace b
+       Data.VectorSpace.InnerSpace a
        ) =>
-  Graphics.Rendering.Diagrams.Bounds.Boundable b where
-    Graphics.Rendering.Diagrams.Bounds.getBounds ::
-      forall b.
+  Graphics.Rendering.Diagrams.Envelope.Enveloped a where
+    Graphics.Rendering.Diagrams.Envelope.getEnvelope ::
+      forall a.
         (
-         Graphics.Rendering.Diagrams.Bounds.Boundable b
+         Graphics.Rendering.Diagrams.Envelope.Enveloped a
          ) =>
-        b ->
-        b
+        a ->
+        a
 
 class (
        Data.VectorSpace.VectorSpace t
@@ -196,6 +229,10 @@
         t
 
 data () =>
+  Graphics.Rendering.Diagrams.Core.NullBackend 
+  deriving ()
+
+data () =>
   Prim' b v =
     forall t. 
     (
@@ -208,6 +245,13 @@
   deriving ()
 
 data () =>
+  Graphics.Rendering.Diagrams.Envelope.LocatedEnvelope v =
+    Graphics.Rendering.Diagrams.Envelope.LocatedEnvelope
+      v
+      v
+  deriving ()
+
+data () =>
   Graphics.Rendering.Diagrams.Names.AName  =
     forall a. 
     (
@@ -257,15 +301,15 @@
   deriving ()
 
 newtype () =>
-  Bounds' v =
-    Bounds
-      Graphics.Rendering.Diagrams.Bounds.appBounds :: (v -> v)
+  Graphics.Rendering.Diagrams.Core.QDiagram b v m =
+    Graphics.Rendering.Diagrams.Core.QD
+      Graphics.Rendering.Diagrams.Core.unQD :: v v m () b v
   deriving ()
 
 newtype () =>
-  Graphics.Rendering.Diagrams.Core.AnnDiagram b v m =
-    Graphics.Rendering.Diagrams.Core.AD
-      Graphics.Rendering.Diagrams.Core.unAD :: v m v b v
+  Envelope' v =
+    Envelope
+      Graphics.Rendering.Diagrams.Envelope.unEnvelope :: (v -> v)
   deriving ()
 
 newtype () =>
@@ -277,18 +321,12 @@
 newtype () =>
   Graphics.Rendering.Diagrams.Names.NameMap v =
     Graphics.Rendering.Diagrams.Names.NameMap
-      Graphics.Rendering.Diagrams.Names.Name [(v, v)]
+      Graphics.Rendering.Diagrams.Names.Name [v]
   deriving ()
 
 newtype () =>
-  Graphics.Rendering.Diagrams.Points.Point v =
-    Graphics.Rendering.Diagrams.Points.P
-      v
-  deriving ()
-
-newtype () =>
-  Graphics.Rendering.Diagrams.Query.Query v m =
-    Graphics.Rendering.Diagrams.Query.Query
+  Query' v m =
+    Query
       Graphics.Rendering.Diagrams.Query.runQuery :: (v -> m)
   deriving ()
 
@@ -299,11 +337,22 @@
   deriving ()
 
 newtype () =>
+  Trace' v =
+    Trace
+      Graphics.Rendering.Diagrams.Trace.appTrace :: (v -> v -> v)
+  deriving ()
+
+newtype () =>
   TransInv' t =
     TransInv
       Graphics.Rendering.Diagrams.Transform.unTransInv :: t
   deriving ()
 
+newtype () =>
+  Data.AffineSpace.Point.Point v =
+    Data.AffineSpace.Point.P
+      v
+  deriving ()
 
 
 
@@ -314,86 +363,37 @@
 
 
 
-infixl 9 Graphics.Rendering.Diagrams.Bounds.appBounds
-Graphics.Rendering.Diagrams.Bounds.appBounds ::
-  forall v.
-    () =>
-    v ->
-    v ->
-    v
 
-infixl 9 Graphics.Rendering.Diagrams.Bounds.boundary
-Graphics.Rendering.Diagrams.Bounds.boundary ::
-  forall a.
-    (
-     Graphics.Rendering.Diagrams.Bounds.Boundable a
-     ) =>
-    a ->
-    a ->
-    a
 
-infixl 9 Graphics.Rendering.Diagrams.Bounds.boundaryFrom
-Graphics.Rendering.Diagrams.Bounds.boundaryFrom ::
-  forall a.
-    (
-     Graphics.Rendering.Diagrams.Bounds.Boundable a
-     ) =>
-    a ->
-    a ->
-    a ->
-    a
-
-infixl 9 Graphics.Rendering.Diagrams.Bounds.boundaryV
-Graphics.Rendering.Diagrams.Bounds.boundaryV ::
-  forall a.
-    (
-     Graphics.Rendering.Diagrams.Bounds.Boundable a
-     ) =>
-    a ->
-    a ->
-    a
-
-infixl 9 Graphics.Rendering.Diagrams.Bounds.diameter
-Graphics.Rendering.Diagrams.Bounds.diameter ::
-  forall a.
-    (
-     Graphics.Rendering.Diagrams.Bounds.Boundable a
-     ) =>
-    a ->
-    a ->
-    a
-
-infixl 9 Graphics.Rendering.Diagrams.Bounds.radius
-Graphics.Rendering.Diagrams.Bounds.radius ::
-  forall a.
-    (
-     Graphics.Rendering.Diagrams.Bounds.Boundable a
-     ) =>
-    a ->
-    a ->
-    a
 
 infixl 6 Graphics.Rendering.Diagrams.Core.atop
 Graphics.Rendering.Diagrams.Core.atop ::
   forall v m b.
     (
-     Data.Monoid.Monoid m
-     ,
-     Graphics.Rendering.Diagrams.Bounds.OrderedField v
+     Graphics.Rendering.Diagrams.Envelope.OrderedField v
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
+     Data.Semigroup.Semigroup m
+     ,
      Data.VectorSpace.InnerSpace v
      ) =>
     b v m ->
     b v m ->
     b v m
 
-infixl 9 Graphics.Rendering.Diagrams.Core.bounds
-Graphics.Rendering.Diagrams.Core.bounds ::
+infixl 9 Graphics.Rendering.Diagrams.Core.clearValue
+Graphics.Rendering.Diagrams.Core.clearValue ::
+  forall b v m.
+    () =>
+    b v m ->
+    b v Data.Monoid.Any
+
+infixl 9 Graphics.Rendering.Diagrams.Core.envelope
+Graphics.Rendering.Diagrams.Core.envelope ::
   forall v b m.
     (
-     Graphics.Rendering.Diagrams.Bounds.OrderedField v
+     Graphics.Rendering.Diagrams.Envelope.OrderedField v
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
@@ -402,20 +402,11 @@
     b v m ->
     v
 
-infixl 9 Graphics.Rendering.Diagrams.Core.clearValue
-Graphics.Rendering.Diagrams.Core.clearValue ::
-  forall b v m.
-    () =>
-    b v m ->
-    b v Data.Monoid.Any
-
 infixl 9 Graphics.Rendering.Diagrams.Core.freeze
 Graphics.Rendering.Diagrams.Core.freeze ::
   forall v b m.
     (
-     Data.Monoid.Monoid m
-     ,
-     Graphics.Rendering.Diagrams.Bounds.OrderedField v
+     Graphics.Rendering.Diagrams.Envelope.OrderedField v
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
@@ -424,13 +415,14 @@
     b v m ->
     b v m
 
-infixl 9 Graphics.Rendering.Diagrams.Core.mkAD
-Graphics.Rendering.Diagrams.Core.mkAD ::
+infixl 9 Graphics.Rendering.Diagrams.Core.mkQD
+Graphics.Rendering.Diagrams.Core.mkQD ::
   forall b v m.
     () =>
     b v ->
     v ->
     v ->
+    v ->
     v m ->
     b v m
 
@@ -438,17 +430,17 @@
 Graphics.Rendering.Diagrams.Core.namePoint ::
   forall v b n m.
     (
-     Data.Monoid.Monoid m
-     ,
-     Graphics.Rendering.Diagrams.Bounds.OrderedField v
+     Graphics.Rendering.Diagrams.Envelope.OrderedField v
      ,
      Graphics.Rendering.Diagrams.Names.IsName n
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
+     Data.Semigroup.Semigroup m
+     ,
      Data.VectorSpace.InnerSpace v
      ) =>
-    (b v m -> (v, v)) ->
+    (b v m -> v) ->
     n ->
     b v m ->
     b v m
@@ -457,14 +449,14 @@
 Graphics.Rendering.Diagrams.Core.named ::
   forall v b n m.
     (
-     Data.Monoid.Monoid m
-     ,
-     Graphics.Rendering.Diagrams.Bounds.OrderedField v
+     Graphics.Rendering.Diagrams.Envelope.OrderedField v
      ,
      Graphics.Rendering.Diagrams.Names.IsName n
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
+     Data.Semigroup.Semigroup m
+     ,
      Data.VectorSpace.InnerSpace v
      ) =>
     n ->
@@ -498,11 +490,9 @@
 
 infixl 9 Graphics.Rendering.Diagrams.Core.prims
 Graphics.Rendering.Diagrams.Core.prims ::
-  forall v m b.
+  forall v b m.
     (
-     Data.Monoid.Monoid m
-     ,
-     Graphics.Rendering.Diagrams.Bounds.OrderedField v
+     Graphics.Rendering.Diagrams.Envelope.OrderedField v
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
@@ -545,22 +535,51 @@
     v ->
     m
 
-infixl 9 Graphics.Rendering.Diagrams.Core.setBounds
-Graphics.Rendering.Diagrams.Core.setBounds ::
+infixl 9 Graphics.Rendering.Diagrams.Core.setEnvelope
+Graphics.Rendering.Diagrams.Core.setEnvelope ::
   forall b v m.
     (
-     Data.Monoid.Monoid m
+     Graphics.Rendering.Diagrams.Envelope.OrderedField v
      ,
-     Graphics.Rendering.Diagrams.Bounds.OrderedField v
+     Graphics.Rendering.Diagrams.Transform.HasLinearMap v
+     ,
+     Data.Monoid.WithSemigroup.Monoid' m
+     ,
+     Data.VectorSpace.InnerSpace v
+     ) =>
+    v ->
+    b v m ->
+    b v m
+
+infixl 9 Graphics.Rendering.Diagrams.Core.setTrace
+Graphics.Rendering.Diagrams.Core.setTrace ::
+  forall b v m.
+    (
+     Graphics.Rendering.Diagrams.Envelope.OrderedField v
      ,
      Graphics.Rendering.Diagrams.Transform.HasLinearMap v
      ,
+     Data.Monoid.WithSemigroup.Monoid' m
+     ,
      Data.VectorSpace.InnerSpace v
      ) =>
     v ->
     b v m ->
     b v m
 
+infixl 9 Graphics.Rendering.Diagrams.Core.trace
+Graphics.Rendering.Diagrams.Core.trace ::
+  forall v b m.
+    (
+     Graphics.Rendering.Diagrams.Transform.HasLinearMap v
+     ,
+     GHC.Classes.Ord v
+     ,
+     Data.VectorSpace.VectorSpace v
+     ) =>
+    b v m ->
+    v
+
 infixl 9 Graphics.Rendering.Diagrams.Core.value
 Graphics.Rendering.Diagrams.Core.value ::
   forall m b v.
@@ -586,7 +605,7 @@
      Data.VectorSpace.InnerSpace v
      ) =>
     n ->
-    ((v, v) -> b v m -> b v m) ->
+    (v -> b v m -> b v m) ->
     b v m ->
     b v m
 
@@ -605,7 +624,7 @@
      Data.VectorSpace.InnerSpace v
      ) =>
     n ->
-    ([(v, v)] -> b v m -> b v m) ->
+    ([v] -> b v m -> b v m) ->
     b v m ->
     b v m
 
@@ -624,10 +643,95 @@
      Data.VectorSpace.InnerSpace v
      ) =>
     [n] ->
-    ([(v, v)] -> b v m -> b v m) ->
+    ([v] -> b v m -> b v m) ->
     b v m ->
     b v m
 
+infixl 9 Graphics.Rendering.Diagrams.Envelope.appEnvelope
+Graphics.Rendering.Diagrams.Envelope.appEnvelope ::
+  forall v.
+    () =>
+    v ->
+    (v -> v)
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.diameter
+Graphics.Rendering.Diagrams.Envelope.diameter ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Envelope.Enveloped a
+     ) =>
+    a ->
+    a ->
+    a
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.envelopeP
+Graphics.Rendering.Diagrams.Envelope.envelopeP ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Envelope.Enveloped a
+     ) =>
+    a ->
+    a ->
+    a
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.envelopeV
+Graphics.Rendering.Diagrams.Envelope.envelopeV ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Envelope.Enveloped a
+     ) =>
+    a ->
+    a ->
+    a
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.inEnvelope
+Graphics.Rendering.Diagrams.Envelope.inEnvelope ::
+  forall v.
+    () =>
+    ((v -> v) -> (v -> v)) ->
+    v ->
+    v
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.locateEnvelope
+Graphics.Rendering.Diagrams.Envelope.locateEnvelope ::
+  forall v.
+    () =>
+    v ->
+    v ->
+    v
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.location
+Graphics.Rendering.Diagrams.Envelope.location ::
+  forall v.
+    () =>
+    v ->
+    v
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.mkEnvelope
+Graphics.Rendering.Diagrams.Envelope.mkEnvelope ::
+  forall v.
+    () =>
+    (v -> v) ->
+    v
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.onEnvelope
+Graphics.Rendering.Diagrams.Envelope.onEnvelope ::
+  forall v.
+    () =>
+    ((v -> v) -> v -> v) ->
+    v ->
+    v
+
+infixl 9 Graphics.Rendering.Diagrams.Envelope.radius
+Graphics.Rendering.Diagrams.Envelope.radius ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Envelope.Enveloped a
+     ) =>
+    a ->
+    a ->
+    a
+
 infixl 9 Graphics.Rendering.Diagrams.HasOrigin.moveOriginBy
 Graphics.Rendering.Diagrams.HasOrigin.moveOriginBy ::
   forall t.
@@ -658,6 +762,19 @@
     t ->
     t
 
+infixl 9 Graphics.Rendering.Diagrams.Juxtapose.juxtaposeDefault
+Graphics.Rendering.Diagrams.Juxtapose.juxtaposeDefault ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Envelope.Enveloped a
+     ,
+     Graphics.Rendering.Diagrams.HasOrigin.HasOrigin a
+     ) =>
+    a ->
+    a ->
+    a ->
+    a
+
 infixr 5 (Graphics.Rendering.Diagrams.Names..>)
 (Graphics.Rendering.Diagrams.Names..>) ::
   forall a1 a2.
@@ -674,11 +791,15 @@
 Graphics.Rendering.Diagrams.Names.fromNames ::
   forall v a.
     (
+     GHC.Float.Floating v
+     ,
      Graphics.Rendering.Diagrams.Names.IsName a
      ,
      GHC.Classes.Ord v
      ,
      Data.AdditiveGroup.AdditiveGroup v
+     ,
+     Data.VectorSpace.InnerSpace v
      ) =>
     [(a, v)] ->
     v
@@ -689,7 +810,7 @@
     (
      Graphics.Rendering.Diagrams.Names.IsName a
      ) =>
-    [(a, (v, v))] ->
+    [(a, v)] ->
     v
 
 infixl 9 Graphics.Rendering.Diagrams.Names.lookupN
@@ -700,7 +821,7 @@
      ) =>
     n ->
     v ->
-    [(v, v)]
+    [v]
 
 infixl 9 Graphics.Rendering.Diagrams.Names.rememberAs
 Graphics.Rendering.Diagrams.Names.rememberAs ::
@@ -711,25 +832,6 @@
     a ->
     v ->
     v ->
-    v ->
-    v
-
-infixl 9 (Graphics.Rendering.Diagrams.Points.*.)
-(Graphics.Rendering.Diagrams.Points.*.) ::
-  forall v.
-    (
-     Data.VectorSpace.VectorSpace v
-     ) =>
-    v ->
-    v ->
-    v
-
-infixl 9 Graphics.Rendering.Diagrams.Points.origin
-Graphics.Rendering.Diagrams.Points.origin ::
-  forall v.
-    (
-     Data.AdditiveGroup.AdditiveGroup v
-     ) =>
     v
 
 infixl 9 Graphics.Rendering.Diagrams.Query.runQuery
@@ -818,6 +920,74 @@
     v ->
     a
 
+infixl 9 Graphics.Rendering.Diagrams.Trace.appTrace
+Graphics.Rendering.Diagrams.Trace.appTrace ::
+  forall v.
+    () =>
+    v ->
+    v ->
+    v ->
+    v
+
+infixl 9 Graphics.Rendering.Diagrams.Trace.inTrace
+Graphics.Rendering.Diagrams.Trace.inTrace ::
+  forall v.
+    () =>
+    ((v -> v -> v) -> v -> v -> v) ->
+    v ->
+    v
+
+infixl 9 Graphics.Rendering.Diagrams.Trace.maxTraceP
+Graphics.Rendering.Diagrams.Trace.maxTraceP ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Trace.Traced a
+     ) =>
+    a ->
+    a ->
+    a ->
+    a
+
+infixl 9 Graphics.Rendering.Diagrams.Trace.maxTraceV
+Graphics.Rendering.Diagrams.Trace.maxTraceV ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Trace.Traced a
+     ) =>
+    a ->
+    a ->
+    a ->
+    a
+
+infixl 9 Graphics.Rendering.Diagrams.Trace.mkTrace
+Graphics.Rendering.Diagrams.Trace.mkTrace ::
+  forall v.
+    () =>
+    (v -> v -> v) ->
+    v
+
+infixl 9 Graphics.Rendering.Diagrams.Trace.traceP
+Graphics.Rendering.Diagrams.Trace.traceP ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Trace.Traced a
+     ) =>
+    a ->
+    a ->
+    a ->
+    a
+
+infixl 9 Graphics.Rendering.Diagrams.Trace.traceV
+Graphics.Rendering.Diagrams.Trace.traceV ::
+  forall a.
+    (
+     Graphics.Rendering.Diagrams.Trace.Traced a
+     ) =>
+    a ->
+    a ->
+    a ->
+    a
+
 infixl 9 (Graphics.Rendering.Diagrams.Transform.<->)
 (Graphics.Rendering.Diagrams.Transform.<->) ::
   forall u v.
@@ -957,15 +1127,21 @@
     t ->
     t
 
-infixl 9 Graphics.Rendering.Diagrams.Util.withLength
-Graphics.Rendering.Diagrams.Util.withLength ::
+infixl 9 (Data.AffineSpace.Point.*.)
+(Data.AffineSpace.Point.*.) ::
   forall v.
     (
-     GHC.Float.Floating v
-     ,
-     Data.VectorSpace.InnerSpace v
+     Data.VectorSpace.VectorSpace v
      ) =>
     v ->
     v ->
     v
 
+infixl 9 Data.AffineSpace.Point.origin
+Data.AffineSpace.Point.origin ::
+  forall v.
+    (
+     Data.AdditiveGroup.AdditiveGroup v
+     ) =>
+    v
+
